projects/poirazi_neuron_model/run_noise_experiment.py

```diff
@@ -0,0 +1,182 @@
+
+        fps = []
+        fns = []
+
+        for trial in range(num_trials):
+
+            successful_initialization = False
+            while not successful_initialization:
+                neuron = Neuron(size = dendrite_length*num_dendrites, num_dendrites = num_dendrites, dendrite_length = dendrite_length, dim = dim, nonlinearity = nonlinearity)
+                data = generate_evenly_distributed_data_sparse(dim = dim, num_active = a, num_samples = num_samples)
+                labels = [1 for i in range(num_samples)]
+
+                neuron.HTM_style_initialize_on_data(data, labels)
+
+                error, fp, fn = get_error(data, labels, [neuron])
+                print "Initialization error is {}, with {} false positives and {} false negatives".format(error, fp, fn)
+                if error == 0:
+                    successful_initialization = True
+                else:
+                    print "Repeating to get a successful initialization"
+
+            apply_noise(data, noise)
+            error, fp, fn = get_error(data, labels, [neuron])
+            fps.append(fp)
+            fns.append(fn)
+            print "Error at noise {} is {}, with {} false positives and {} false negatives".format(noise, error, fp, fn)
+
+        with open("noise_FN_{}.txt".format(theta), "a") as f:
+            f.write(str(noise) + ", " + str(numpy.sum(fns)) + ", " + str(num_trials*num_samples) + "\n")
+
+
+
+def run_pm_style_noise_experiment(num_neurons = 1,
+                         num_neg_neurons = 1,
+                         a = 128,
+                         dim = 6000,
+                         test_noise_levels = range(30, 100, 5),
+                         num_samples = 1000,
+                         num_dendrites = 500,
+                         dendrite_length = 30,
+                         num_replacement = 25,
+                         training_iterations = 10000,
+                         new_style_training = True,
+                         num_trials = 200,
+                         nonlinearity = power_nonlinearity(10)):
+    """
+    A version of the above test, but using two P&M style competing neurons with
+    nonlinearity l(x) = x^2.  As the test is now symmetric, with a false
+    positive for one neuron being a false negative for the other, we only track
+    overall error.  There is no "theta", as identification is merely dependent
+    on which neuron in the pair responds more strongly.  This is most likely
+    not biologically plausible, especially at high noise values.
+
+    Training is done via HTM-style initialization.  In the event that the init
+    fails to produce an error rate of 0 without noise (which anecdotally never
+    occurs), we simple reinitialize.
+
+    Results are saved to the file pm_noise_FN.txt.
+    """
+    for noise in test_noise_levels:
+
+        fps = []
+        fns = []
+
+        for trial in range(num_trials):
+
+            successful_initialization = False
```

Comment:
I'm not sure the logic here would actually work out, given that I only want to re-initialize if there's a problem.